cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+    {
+        Log(LOG_LEVEL_INFO, "Can not update installed packages cache after "
+            "package installation");
+        return PROMISE_RESULT_FAIL;
+    }
+    
+    if (!UpdateCache(policy_data->package_options, wrapper,
+                     UPDATE_TYPE_LOCAL_UPDATES))
+    {
+        Log(LOG_LEVEL_INFO, "Can not update available updates cache after "
+            "package installation");
+        return PROMISE_RESULT_FAIL;
+    }
+
+    int is_in_cache = IsPackageInCache(NULL, wrapper, package_info->name,
+                                       package_info->version,
+                                       package_info->arch);
+    if (is_in_cache == 1)
+    {
+        return action_type == NEW_PACKAGE_ACTION_PRESENT ? 
+            PROMISE_RESULT_CHANGE : PROMISE_RESULT_FAIL;
+    }
+    else if (is_in_cache == 0)
+    {
+        return action_type == NEW_PACKAGE_ACTION_PRESENT ? 
+            PROMISE_RESULT_FAIL : PROMISE_RESULT_CHANGE;
+    }
+    else
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while reading installed "
+            "packages cache.");
+        return PROMISE_RESULT_FAIL;
+    }
+}
+
+PromiseResult RemovePackage(const char *name, Rlist* options, 
+                            const char *version, const char *architecture,
+                            const PackageModuleWrapper *wrapper)
+{
+    Log(LOG_LEVEL_DEBUG, "Removing package '%s'", name);
+             
+    char *options_str = ParseOptions(options);
+    char *ver = version ? 
+        StringFormat("Version=%s\n", version) : NULL;
+    char *arch = architecture ? 
+        StringFormat("Architecture=%s\n", architecture) : NULL;
+    char *request = StringFormat("%sName=%s\n%s%s",
+            options_str, name, ver ? ver : "", arch ? arch : "");
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    Rlist *error_message = NULL;
+    if (ReadWriteDataToPackageScript("remove", request, 
+            &error_message, wrapper) != 0)
+    {
+        Log(LOG_LEVEL_INFO, "Some error occurred while communicating with "
+            "package module while removing package.");
+        res = PROMISE_RESULT_FAIL;
+    }
+    if (error_message)
+    {
+        ParseAndLogErrorMessage(error_message);
+        res = PROMISE_RESULT_FAIL;
+        RlistDestroy(error_message);
+    }
+    
+    free(request);
+    free(options_str);
+    free(ver);
+    free(arch);
+    
+    /* We assume that at this point package is removed correctly. */
+    return res;    
+}
+
+
+static PromiseResult InstallPackageGeneric(Rlist *options, 
+        PackageType type, const char *packages_list_formatted,
+        const PackageModuleWrapper *wrapper)
+{
+    char *options_str = ParseOptions(options);
+    char *request = StringFormat("%s%s", options_str, packages_list_formatted);
+    
+    PromiseResult res = PROMISE_RESULT_CHANGE;
+    
+    const char *package_install_command = NULL;
+    if (type == PACKAGE_TYPE_FILE)
```

Comment:
Always for more than 3 checks. I suppose in simpler cases (or even when more checks are done) compiled code will be the same. For readability I think 3 cases is quite OK to use ```if-else if```.